service/matching/matching_engine.go

```diff
@@ -370,61 +370,64 @@ func (e *matchingEngineImpl) getTaskQueuePartitionManager(
 	partition tqid.Partition,
 	create bool,
 	loadCause loadCause,
-) (taskQueuePartitionManager, error) {
-	pm, err := e.getTaskQueuePartitionManagerNoWait(partition, create, loadCause)
-	if err != nil || pm == nil {
-		return nil, err
-	}
-	if err = pm.WaitUntilInitialized(ctx); err != nil {
-		e.unloadTaskQueuePartition(pm, unloadCauseInitError)
-		return nil, err
-	}
-	return pm, nil
-}
-
-// Returns taskQueuePartitionManager for a task queue. If not already cached, and create is true, tries
-// to get new range from DB and create one. This does not block for the task queue to be initialized.
-func (e *matchingEngineImpl) getTaskQueuePartitionManagerNoWait(
-	partition tqid.Partition,
-	create bool,
-	loadCause loadCause,
-) (taskQueuePartitionManager, error) {
+) (taskQueuePartitionManager, bool, error) {
 	key := partition.Key()
+
 	e.partitionsLock.RLock()
 	pm, ok := e.partitions[key]
 	e.partitionsLock.RUnlock()
-	if !ok {
-		if !create {
-			return nil, nil
-		}
 
-		// If it gets here, write lock and check again in case a task queue is created between the two locks
-		e.partitionsLock.Lock()
-		pm, ok = e.partitions[key]
-		if !ok {
-			namespaceEntry, err := e.namespaceRegistry.GetNamespaceByID(partition.NamespaceId())
-			if err != nil {
-				e.partitionsLock.Unlock()
-				return nil, err
-			}
-			nsName := namespaceEntry.Name()
-			tqConfig := newTaskQueueConfig(partition.TaskQueue(), e.config, nsName)
-			tqConfig.loadCause = loadCause
-			userDataManager := newUserDataManager(e.taskManager, e.matchingRawClient, partition, tqConfig, e.logger, e.namespaceRegistry)
-			pm, err = newTaskQueuePartitionManager(e, namespaceEntry, partition, tqConfig, userDataManager)
-			if err != nil {
-				e.partitionsLock.Unlock()
-				return nil, err
-			}
-			e.partitions[key] = pm
-		}
+	if ok {
+		return pm, false, nil
```

Comment:
Actually there's one more problem here, and I'm embarrassed to say I didn't notice it until someone else pointed it out:
The pm that you read from the map may be initialized yet. E.g. if another thread just inserted it and that one is still blocked on WaitUntilInitialized. The correct thing to do is always call WaitUntilInitialized, whether it came from the map or was just constructed. The previous organization (with the NoWait call) did that. You could also maybe do it with a defer.

Reply:
David and I addressed this by pair programming getTaskQueuePartitionManager